p2p/src/authenticated/actors/tracker/record.rs

```diff
@@ -138,81 +146,164 @@ impl<C: Verifier> Record<C> {
+    pub fn reserve(&mut self) -> bool {
+        if matches!(self.address, Address::Blocked | Address::Myself(_)) {
             return false;
         }
-        match self.address {
-            Address::Blocked | Address::Unknown | Address::Discovered(_) => true,
-            Address::Myself(_) | Address::Bootstrapper(_) | Address::Persistent(_) => false,
+        if matches!(self.status, Status::Inert) {
+            self.status = Status::Reserved;
+            return true;
+        }
+        false
+    }
+
+    /// Marks the peer as connected.
+    ///
+    /// The peer must have the status [`Status::Reserved`].
+    pub fn connect(&mut self) {
+        assert!(matches!(self.status, Status::Reserved));
+        self.status = Status::Active;
+    }
+
+    /// Releases any reservation on the peer.
+    pub fn release(&mut self) {
+        assert!(self.status != Status::Inert, "Cannot release an Inert peer");
+        self.status = Status::Inert;
+    }
+
+    /// Indicate that there was a dial failure for this peer using the given `socket`, which is
+    /// checked against the existing record to ensure that we correctly attribute the failure.
+    pub fn dial_failure(&mut self, socket: SocketAddr) {
+        if let Address::Discovered(info, fails) = &mut self.address {
+            if info.socket == socket {
+                *fails += 1;
+            }
+        }
+    }
+
+    /// Indicate that a dial succeeded for this peer.
+    ///
+    /// Due to race conditions, it's possible that we connected using a socket that is now ejected
+    /// from the record. However, in this case, the record would already have the `fails` set to 0,
+    /// so we can avoid checking against the socket.
+    pub fn dial_success(&mut self) {
+        if let Address::Discovered(_, fails) = &mut self.address {
+            *fails = 0;
         }
     }
 
     // ---------- Getters ----------
 
-    /// Check if the peer is blocked.
+    /// Returns `true` if the record is blocked.
     pub fn blocked(&self) -> bool {
         matches!(self.address, Address::Blocked)
     }
 
-    /// Returns true if we don't need additional [`PeerInfo`] about this peer.
+    /// Returns `true` if the record is dialable.
     ///
-    /// Similar to `peer_info().is_some()`, but also include the `Blocked` case since we don't
-    /// want to track blocked peers despite not recording their information.
-    pub fn discovered(&self) -> bool {
-        match self.address {
-            Address::Unknown | Address::Bootstrapper(_) => false,
-            Address::Myself(_)
-            | Address::Blocked
-            | Address::Discovered(_)
-            | Address::Persistent(_) => true,
-        }
+    /// A record is dialable if:
+    /// - We have the socket address of the peer
+    /// - It is not ourselves
+    /// - We are not already connected
+    pub fn dialable(&self) -> bool {
+        matches!(self.status, Status::Inert | Status::Reserved)
```

Comment:
nit: this more or less "pessimistically assumes" that we won't connect on the current reservation we have (and should plan to try again or else will have to wait for the next query refresh).

I think we should add a comment explaining a bit more above `matches`

Reply:
This is sort of best-effort anyway, but maybe we should just ignore `Reserved` since even if we return those values, they are likely to be rate-limited. What do you think?

Reply:
I think whatever we can do to reduce the size of the dialable array is best. The only way it could be reserved is if we are actively dialing (or they are dialing us). Seems ok to not include.